tests/k8st/tests/test_bgp_advert.py

```diff
@@ -143,6 +228,112 @@ def assert_ecmp_routes(self, dst, via=["10.192.0.3", "10.192.0.4"]):
+metadata:
+  name: nginx-rr
+  namespace: bgp-test
+  labels:
+    app: nginx
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: nginx
+      run: nginx-rr
+  template:
+    metadata:
+      labels:
+        app: nginx
+        run: nginx-rr
+    spec:
+      containers:
+      - name: nginx-rr
+        image: nginx:1.7.9
+        ports:
+        - containerPort: 80
+      nodeSelector:
+        beta.kubernetes.io/os: linux
+        kubernetes.io/hostname: kube-node-1
+---
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: nginx-rr
+  namespace: bgp-test
+  labels:
+    app: nginx
+    run: nginx-rr
+spec:
+  ports:
+  - port: 80
+    targetPort: 80
+  selector:
+    app: nginx
+    run: nginx-rr
+  type: NodePort
+  externalTrafficPolicy: Local
+EOF
+""")
+
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get nodes -o yaml")
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get bgppeers -o yaml")
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get bgpconfigs -o yaml")
+
+        # Update the node-2 to behave as a route-reflector
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get node kube-node-2 -o json > /home/node.json")
+        run("sed -i '11 a \ \ \ \ \ \ \"i-am-a-route-reflector\": \"true\",' /home/node.json")
+        run("sed -i '21 a \ \ \ \ \ \ \"routeReflectorClusterID\": \"224.0.0.1\",' /home/node.json")
+        run("kubectl cp /home/node.json kube-system/calicoctl:/home/node.json")
+        run("kubectl exec -i -n kube-system calicoctl -- cat /home/node.json")
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl apply -f /home/node.json")
+        run("kubectl exec -i -n kube-system calicoctl -- /calicoctl get node kube-node-2 -o yaml")
+
+        # Disable node-to-node mesh and configure bgp peering
+        # between node-1 and RR and also between external node and RR
+        run("""kubectl exec -i -n kube-system calicoctl -- /calicoctl apply -f - << EOF
+apiVersion: projectcalico.org/v3
+items:
+- apiVersion: projectcalico.org/v3
+  kind: BGPConfiguration
+  metadata: {name: default}
+  spec:
+    nodeToNodeMeshEnabled: false
+    asNumber: 64512
+kind: BGPConfigurationList
+metadata:
+  resourceVersion: 139
+EOF
+""")
+        run("""kubectl exec -i -n kube-system calicoctl -- /calicoctl apply -f - << EOF
+apiVersion: projectcalico.org/v3
+items:
+- apiVersion: projectcalico.org/v3
+  kind: BGPPeer
+  metadata: {name: kube-node-1}
+  spec:
+    node: kube-node-1
+    peerIP: 10.192.0.4
+    asNumber: 64512
+kind: BGPPeerList
+EOF
+""")
+        run("kubectl get endpoints nginx-rr -n bgp-test -o yaml > /home/endpoint.yaml")
+        svcRoute = ""
+        with open("/home/endpoint.yaml", 'r') as stream:
+            content = yaml.load(stream)
+            for k,v in content.items():
+                if "clusterIP" in v:
+                    svcRoute = k
+        self.assertIn(svcRoute, self.get_routes())
```

Comment:
I think we will need to allow some retries here, for the configuration and BGP updates to flow through.  So I think we need:
```
retry_until_success(lambda: self.assertIn(svcRoute, self.get_routes()))
```
That will try up to 10 times, with 1s interval.  If that isn't enough, we can use the `retries` and `wait_time` kwargs, for example:
```
retry_until_success(lambda: self.assertIn(svcRoute, self.get_routes()),
                                retries=12, wait_time=3)
```

Reply:
acknowledged